### PR TITLE
HIVE-25346: cleanTxnToWriteIdTable breaks SNAPSHOT isolation

### PR DESCRIPTION
**Pull request by zchovan**
initial test fixes
Change-Id: Ieb4f922d1e1957538cbeda2d410a167d18993724

### CODE REVIEW

**Comment by zchovan** [comment]
retest

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 1]
```diff
@@ -531,7 +531,7 @@ public void cleanTxnToWriteIdTable() throws MetaException {
         String s = "SELECT MIN(\"RES\".\"ID\") AS \"ID\" FROM (" +
             "SELECT MAX(\"TXN_ID\") + 1 AS \"ID\" FROM \"TXNS\" " +
             "UNION " +
-            "SELECT MIN(\"WS_COMMIT_ID\") AS \"ID\" FROM \"WRITE_SET\" " +
+            "SELECT MIN(\"WS_TXNID\") AS \"ID\" FROM \"WRITE_SET\" " +
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:534]: @zchovan  are you sure that is an equivalent change? @lcspinter what do you think?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 2]
```diff
@@ -1442,7 +1443,9 @@ public void commitTxn(CommitTxnRequest rqst)
         }
 
         String conflictSQLSuffix = "FROM \"TXN_COMPONENTS\" WHERE \"TC_TXNID\"=" + txnid + " AND \"TC_OPERATION_TYPE\" IN (" +
-                OperationType.UPDATE + "," + OperationType.DELETE + ")";
+                OperationType.UPDATE + "," + OperationType.DELETE  + "," + OperationType.INSERT + ")";
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1446]: Why is this needed? Looks like you are handling Inserts in a separate if block.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 3]
```diff
@@ -1442,7 +1443,9 @@ public void commitTxn(CommitTxnRequest rqst)
...
+        String insertFilter = "FROM \"TXN_COMPONENTS\" WHERE  \"TC_TXNID\"=" + txnid +
+            " AND \"TC_OPERATION_TYPE\" IN (" + OperationType.INSERT + ")";
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1448]: Why not simply '=' instead of IN

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 4]
```diff
@@ -1513,6 +1516,11 @@ public void commitTxn(CommitTxnRequest rqst)
         } else if (txnType.get() == TxnType.COMPACTION) {
           acquireTxnLock(stmt, false);
           commitId = getHighWaterMark(stmt);
+        } else if (txnType.get() != TxnType.READ_ONLY && isInsert(stmt, insertFilter)) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1519]: We should ignore repl as well. Please extract "txnType.get() != TxnType.READ_ONLY && !rqst.isSetReplPolicy()" nand handle inserts and updates there.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 5]
```diff
@@ -1608,6 +1616,14 @@ private boolean isUpdateOrDelete(Statement stmt, String conflictSQLSuffix) throw
     }
   }
 
+  private boolean isInsert(Statement stmt, String insertFilter) throws SQLException, MetaException {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1619]: this method has a specific purpose, why not define filter inside?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 6]
```diff
@@ -1513,6 +1516,11 @@ public void commitTxn(CommitTxnRequest rqst)
         } else if (txnType.get() == TxnType.COMPACTION) {
           acquireTxnLock(stmt, false);
           commitId = getHighWaterMark(stmt);
+        } else if (txnType.get() != TxnType.READ_ONLY && isInsert(stmt, insertFilter)) {
+          stmt.executeUpdate("INSERT INTO \"WRITE_SET\" (\"WS_DATABASE\", \"WS_TABLE\", \"WS_PARTITION\", \"WS_TXNID\", \"WS_COMMIT_ID\", \"WS_OPERATION_TYPE\")" +
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1520]: do we have an index on TC_OPERATION_TYPE?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 7]
```diff
@@ -1513,6 +1516,11 @@ public void commitTxn(CommitTxnRequest rqst)
...
+              " SELECT DISTINCT \"TC_DATABASE\", \"TC_TABLE\", \"TC_PARTITION\", \"TC_TXNID\", " + tempCommitId + ", \"TC_OPERATION_TYPE\" " + insertFilter);
+          acquireTxnLock(stmt, false);
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1522]: Why do you need to acquire locks here?

**Review by deniskuzZ — CHANGES_REQUESTED** [review]
@zchovan thank you for the patch, added a number of review comments

**Comment by deniskuzZ** [comment]
Would be great if you could run the HMS benchmark and see if that has affected the commit step performace. See if the index on opType improves the. situation.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 7 above)
Comment (zchovan) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1522]: it's required for later when ` updateWSCommitIdAndCleanUpMetadata(stmt, txnid, txnType.get(), commitId, tempCommitId);` is called

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 6 above)
Comment (zchovan) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: no, we don't

**Comment by zchovan** [comment]
> Would be great if you could run the HMS benchmark and see if that has affected the commit step performace. See if the index on opType improves the. situation.

There is no benchmark currently for the commitTxn() call.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 6 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: how would it affect the performance?

**Comment by deniskuzZ** [comment]
> > Would be great if you could run the HMS benchmark and see if that has affected the commit step performance. See if the index on opType improves the situation.
> 
> There is no benchmark currently for the commitTxn() call.

Could we create one or at least perform similar to https://issues.apache.org/jira/browse/HIVE-23104?focusedCommentId=17083005&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17083005 test?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 6 above)
Comment (zchovan) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java]: The row count would most likely be higher in the WRITE_SET table, so selects could possibly see slower response times.

**Comment by zchovan** [comment]
> > > Would be great if you could run the HMS benchmark and see if that has affected the commit step performance. See if the index on opType improves the situation.
> > 
> > 
> > There is no benchmark currently for the commitTxn() call.
> 
> Could we create one or at least perform similar to https://issues.apache.org/jira/browse/HIVE-23104?focusedCommentId=17083005&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17083005 test?

Based on the linked comment it is not clear to me what exactly was tested and how or what tools were used for it or what the environment the tests were run in. I found no mention of the above details in the jira comments.

**Comment by zchovan** [comment]
@deniskuzZ I did some benchmarks on commitTxns, please find the results below:


`
Before HIVE-25346

Result "org.apache.hadoop.hive.metastore.tools.ACIDBenchmarks.TestOpenTxn.openTxn":
  N = 100
  mean =     58.136 ±(99.9%) 1.757 ms/op

  Histogram, ms/op:
    [40.000, 42.500) = 0
    [42.500, 45.000) = 0
    [45.000, 47.500) = 0
    [47.500, 50.000) = 2
    [50.000, 52.500) = 11
    [52.500, 55.000) = 14
    [55.000, 57.500) = 20
    [57.500, 60.000) = 26
    [60.000, 62.500) = 13
    [62.500, 65.000) = 7
    [65.000, 67.500) = 1
    [67.500, 70.000) = 3
    [70.000, 72.500) = 0
    [72.500, 75.000) = 1
    [75.000, 77.500) = 1

  Percentiles, ms/op:
      p(0.0000) =     49.111 ms/op
     p(50.0000) =     57.856 ms/op
     p(90.0000) =     63.579 ms/op
     p(95.0000) =     67.901 ms/op
     p(99.0000) =     79.537 ms/op
     p(99.9000) =     79.567 ms/op
     p(99.9900) =     79.567 ms/op
     p(99.9990) =     79.567 ms/op
     p(99.9999) =     79.567 ms/op
    p(100.0000) =     79.567 ms/op


After HIVE-25346


Result "org.apache.hadoop.hive.metastore.tools.ACIDBenchmarks.TestOpenTxn.openTxn":
  N = 100
  mean =     51.015 ±(99.9%) 2.947 ms/op

  Histogram, ms/op:
    [30.000, 35.000) = 0
    [35.000, 40.000) = 3
    [40.000, 45.000) = 26
    [45.000, 50.000) = 27
    [50.000, 55.000) = 17
    [55.000, 60.000) = 11
    [60.000, 65.000) = 9
    [65.000, 70.000) = 4
    [70.000, 75.000) = 1
    [75.000, 80.000) = 1
    [80.000, 85.000) = 1

  Percentiles, ms/op:
      p(0.0000) =     38.385 ms/op
     p(50.0000) =     48.248 ms/op
     p(90.0000) =     64.176 ms/op
     p(95.0000) =     68.280 ms/op
     p(99.0000) =     81.816 ms/op
     p(99.9000) =     81.838 ms/op
     p(99.9900) =     81.838 ms/op
     p(99.9990) =     81.838 ms/op
     p(99.9999) =     81.838 ms/op
    p(100.0000) =     81.838 ms/op

`

**Comment by deniskuzZ** [comment]
@zchovan thank you for the benchmarks! 
Last thing, to be on a safe side, could you please record INSERTS only if min_history is not available + use old logic for writeId removal based on min_history if available. @pvary, what do you think?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java (hunk 1 above)
Comment (pvary) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:535]: If my memory serves me well, `WS_COMMIT_ID` is the `NEXT_TXN_ID` at the time of the commit, and the `WS_TXNID` is the id of the given transaction

Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:535]: that is correct. However, I don't understand how this is supposed to replace MIN_HISTORY table. Original code:
````
long minUncommittedTxnId = findMinOpenTxnGLB(stmt);
// If there are aborted txns, then the minimum aborted txnid could be the min_uncommitted_txnid
// if lesser than both NEXT_TXN_ID.ntxn_next and min(MIN_HISTORY_LEVEL .mhl_min_open_txnid).
String s = "SELECT MIN(\"TXN_ID\") FROM \"TXNS\" WHERE \"TXN_STATE\" = " + quoteChar(TXN_ABORTED);
minUncommittedTxnId = Math.min(minAbortedTxnId, minUncommittedTxnId);
```

Comment (pvary) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:535]: This was the original code, so we might want to check `findMinOpenTxnGLB`:
```
        // First need to find the min_uncommitted_txnid which is currently seen by any open transactions.
        // If there are no txns which are currently open or aborted in the system, then current value of
        // NEXT_TXN_ID.ntxn_next could be min_uncommitted_txnid.
        long minUncommittedTxnId = findMinOpenTxnGLB(stmt);

        // If there are aborted txns, then the minimum aborted txnid could be the min_uncommitted_txnid
        // if lesser than both NEXT_TXN_ID.ntxn_next and min(MIN_HISTORY_LEVEL .mhl_min_open_txnid).
        String s = "SELECT MIN(\"TXN_ID\") FROM \"TXNS\" WHERE \"TXN_STATE\" = " + TxnStatus.ABORTED;
        LOG.debug("Going to execute query <" + s + ">");
        rs = stmt.executeQuery(s);
        if (rs.next()) {
          long minAbortedTxnId = rs.getLong(1);
          if (minAbortedTxnId > 0) {
            minUncommittedTxnId = Math.min(minAbortedTxnId, minUncommittedTxnId);
          }
        }
```

Comment (pvary) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:535]: This was the `findMinOpenTxnGLB`
```
  /**
   * See doc at {@link TxnStore#findMinOpenTxnId()}
   * Note that {@link #openTxns(OpenTxnRequest)} makes update of NEXT_TXN and MIN_HISTORY_LEVEL
   * a single atomic operation (and no one else should update these tables except the cleaner
   * which deletes rows from MIN_HISTORY_LEVEL which can only allow minOpenTxn to move higher)
   */
  private long findMinOpenTxnGLB(Statement stmt) throws MetaException, SQLException {
    String s = "SELECT \"NTXN_NEXT\" FROM \"NEXT_TXN_ID\"";
    LOG.debug("Going to execute query <" + s + ">");
    ResultSet rs = stmt.executeQuery(s);
    if (!rs.next()) {
      throw new MetaException("Transaction tables not properly " +
          "initialized, no record found in next_txn_id");
    }
    long hwm = rs.getLong(1);
    s = "SELECT MIN(\"MHL_MIN_OPEN_TXNID\") FROM \"MIN_HISTORY_LEVEL\"";
    LOG.debug("Going to execute query <" + s + ">");
    rs = stmt.executeQuery(s);
    rs.next();
    long minOpenTxnId = rs.getLong(1);
    if(rs.wasNull()) {
      return hwm;
    }
    //since generating new txnid uses select for update on single row in NEXT_TXN_ID
    assert hwm >= minOpenTxnId : "(hwm, minOpenTxnId)=(" + hwm + "," + minOpenTxnId + ")";
    return minOpenTxnId;
  }
```

Comment (pvary) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:535]: So it think the theory is this:
- Original: Minimum open transactionId which was open when any of the currently executing transactions started.
- New: Minimum open transactionId which was open when the last write committed.

So originally we removed lines where the transaction older of any open/aborted transactions, and even the running transactions were not needed to see the older changes, but this was based on the starting time of the transaction.

With the new change we are only keeping the `WRITE_SET` based on the transactionId when the actual write was committed (not when the actual write was started), as the previous queries should not read the non-committed folder/data anyway

Question: What happens with the `WRITE_SET` when a compaction happened on this table (we are relying here on the fact that we are the only ones removing lines from this table)

Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:535]: ````With the new change we are only keeping the WRITE_SET based on the transactionId when the actual write was committed (not when the actual write was started), as the previous queries should not read the non-committed folder/data anyway````
that leads to the violation of the SNAPSHOT isolation, see `testUpdateSnapshotIsolation test

Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:535]: ````What happens with the WRITE_SET```` - nothing, it's only cleaned by the AcidHouseKeeperThread:
````
performTask(txnHandler::performWriteSetGC, "Cleaning obsolete write set entries");
performTask(txnHandler::cleanTxnToWriteIdTable, "Cleaning obsolete TXN_TO_WRITE_ID entries");
````

Comment (pvary) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:535]: +1

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 8]
```diff
@@ -1446,70 +1446,75 @@ public void commitTxn(CommitTxnRequest rqst)
...
+                    throw new IllegalStateException(msg + " FAILED!");
+                  }
+                  dbConn.commit();
+                  throw new TxnAbortedException(msg);
                 }
-                dbConn.commit();
-                throw new TxnAbortedException(msg);
               }
             }
+          } else if (isInsert(stmt, txnid)) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1512]: You don't have to check for `insert` as there is no other alternative.

**Comment by kgyrtkirk** [comment]
merged as #2716
